Flaky Test: oshi.hardware.CentralProcessorTest.testFrequencies()

From [CI](https://github.com/oshi/oshi/actions/runs/7228489323):
```
    => java.lang.AssertionError: Logical processor frequency should be at most it's max frequency
Expected: is a value less than or equal to <3241340000L>
     but: <3242130000L> was greater than <3241340000L>
       org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
       oshi.hardware.CentralProcessorTest.testFrequencies(CentralProcessorTest.java:68)
```
and
```
    => java.lang.AssertionError: Logical processor frequency should be at most it's max frequency
Expected: is a value less than or equal to <3243848000L>
     but: <3244070000L> was greater than <3243848000L>
       org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
       oshi.hardware.CentralProcessorTest.testFrequencies(CentralProcessorTest.java:68)
```

--- a/oshi-core/src/main/java/oshi/hardware/common/AbstractCentralProcessor.java
+++ b/oshi-core/src/main/java/oshi/hardware/common/AbstractCentralProcessor.java
@@ -41,7 +41,8 @@ public abstract class AbstractCentralProcessor implements CentralProcessor {
 
     private final Supplier<ProcessorIdentifier> cpuid = memoize(this::queryProcessorId);
     private final Supplier<Long> maxFreq = memoize(this::queryMaxFreq, defaultExpiration());
-    private final Supplier<long[]> currentFreq = memoize(this::queryCurrentFreq, defaultExpiration());
+    // Max often iterates current, intentionally making it shorter to re-memoize current
+    private final Supplier<long[]> currentFreq = memoize(this::queryCurrentFreq, defaultExpiration() / 2L);
     private final Supplier<Long> contextSwitches = memoize(this::queryContextSwitches, defaultExpiration());
     private final Supplier<Long> interrupts = memoize(this::queryInterrupts, defaultExpiration());
 

--- a/oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxCentralProcessor.java
+++ b/oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxCentralProcessor.java
@@ -490,13 +490,8 @@ final class LinuxCentralProcessor extends AbstractCentralProcessor {
 
     @Override
     public long queryMaxFreq() {
-        long max = Arrays.stream(this.getCurrentFreq()).max().orElse(-1L);
-        // Max of current freq, if populated, is in units of Hz, convert to kHz
-        if (max > 0) {
-            max /= 1000L;
-        }
-        // Iterating CPUs only gets the existing policy, so we need to iterate the
-        // policy directories to find the system-wide policy max
+        long policyMax = -1L;
+        // Iterate the policy directories to find the system-wide policy max
         UdevContext udev = Udev.INSTANCE.udev_new();
         try {
             UdevEnumerate enumerate = udev.enumerateNew();
@@ -508,7 +503,7 @@ final class LinuxCentralProcessor extends AbstractCentralProcessor {
                 UdevListEntry entry = enumerate.getListEntry();
                 if (entry != null) {
                     String syspath = entry.getName(); // /sys/devices/system/cpu/cpu0
-                    String cpuFreqPath = syspath.substring(0, syspath.lastIndexOf(File.separatorChar)) + "/cpuFreq";
+                    String cpuFreqPath = syspath.substring(0, syspath.lastIndexOf(File.separatorChar)) + "/cpufreq";
                     String policyPrefix = cpuFreqPath + "/policy";
                     try (Stream<Path> path = Files.list(Paths.get(cpuFreqPath))) {
                         Optional<Long> maxPolicy = path.filter(p -> p.toString().startsWith(policyPrefix)).map(p -> {
@@ -518,8 +513,9 @@ final class LinuxCentralProcessor extends AbstractCentralProcessor {
                             }
                             return freq;
                         }).max(Long::compare);
-                        if (maxPolicy.isPresent() && max < maxPolicy.get()) {
-                            max = maxPolicy.get();
+                        if (maxPolicy.isPresent()) {
+                            // Value is in kHz
+                            policyMax = maxPolicy.get() * 1000L;
                         }
                     } catch (IOException e) {
                         // ignore
@@ -531,15 +527,11 @@ final class LinuxCentralProcessor extends AbstractCentralProcessor {
         } finally {
             udev.unref();
         }
-        if (max == 0L) {
-            return -1L;
-        }
-        // If successful, value is in KHz.
-        max *= 1000L;
-        // Cpufreq result assumes intel pstates and is unreliable for AMD processors.
         // Check lshw as a backup
         long lshwMax = Lshw.queryCpuCapacity();
-        return lshwMax > max ? lshwMax : max;
+        // And get the highest of existing current frequencies
+        return LongStream.concat(LongStream.of(policyMax, lshwMax), Arrays.stream(this.getCurrentFreq())).max()
+                .orElse(-1L);
     }
 
     @Override

--- a/oshi-core/src/test/java/oshi/hardware/CentralProcessorTest.java
+++ b/oshi-core/src/test/java/oshi/hardware/CentralProcessorTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016-2022 The OSHI Project Contributors
+ * Copyright 2016-2023 The OSHI Project Contributors
  * SPDX-License-Identifier: MIT
  */
 package oshi.hardware;
@@ -65,7 +65,7 @@ class CentralProcessorTest {
                 p.getLogicalProcessorCount(), is(curr.length));
         if (max >= 0) {
             for (long freq : curr) {
-                assertThat("Logical processor frequency should be at most it's max frequency", freq,
+                assertThat("Logical processor frequency should be at most its max frequency", freq,
                         is(lessThanOrEqualTo(max)));
             }
         }

